src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -122,26 +138,47 @@ private IModelBinder CreateBinderCore(DefaultModelBinderProviderContext provider
                 }
             }
 
-            if (result == null && stack.Count > 1)
+            // If the DelegatingBinder was created, then it means we recursed. Hook it up to the 'real' binder.
+            var delegatingBinder = collection[key] as PlaceholderBinder;
+            if (delegatingBinder != null)
             {
-                // Use a no-op binder if we're below the top level. At the top level, we throw.
-                result = NoOpBinder.Instance;
+                // It's also possible that user code called into `CreateBinder` but then returned null, we don't
+                // want to create something that will null-ref later so just hook this up to the no-op binder.
+                delegatingBinder.Inner = result ?? NoOpBinder.Instance;
             }
 
-            // "pop"
-            Debug.Assert(stack.Count > 0);
-            var delegatingBinder = stack[stack.Count - 1].Value;
-            stack.RemoveAt(stack.Count - 1);
-
-            // If the DelegatingBinder was created, then it means we recursed. Hook it up to the 'real' binder.
-            if (delegatingBinder != null)
+            if (result != null)
             {
-                delegatingBinder.Inner = result;
+                collection[key] = result;
             }
 
             return result;
         }
 
+        private void AddToCache(ModelMetadata metadata, object cacheToken, IModelBinder binder)
+        {
+            Debug.Assert(metadata != null);
+            Debug.Assert(binder != null);
+
+            if (cacheToken == null)
```

Comment:
Why support the `cacheToken == null` case at all? Not used for inner nodes and makes even less sense for top-level binders.